libs/database/src/publish.rs

```diff
@@ -0,0 +1,273 @@
+use app_error::AppError;
+use database_entity::dto::{PublishCollabItem, PublishInfo};
+use sqlx::{Executor, PgPool, Postgres};
+use uuid::Uuid;
+
+pub async fn select_user_is_collab_publisher_for_all_views(
+  pg_pool: &PgPool,
+  user_uuid: &Uuid,
+  workspace_uuid: &Uuid,
+  view_ids: &[Uuid],
+) -> Result<bool, AppError> {
+  let count = sqlx::query_scalar!(
+    r#"
+      SELECT COUNT(*)
+      FROM af_published_collab
+      WHERE workspace_id = $1
+        AND view_id = ANY($2)
+        AND published_by = (SELECT uid FROM af_user WHERE uuid = $3)
+    "#,
+    workspace_uuid,
+    view_ids,
+    user_uuid,
+  )
+  .fetch_one(pg_pool)
+  .await?;
+
+  match count {
+    Some(c) => Ok(c == view_ids.len() as i64),
+    None => Ok(false),
+  }
+}
+
+#[inline]
+pub async fn select_workspace_publish_namespace_exists<'a, E: Executor<'a, Database = Postgres>>(
+  executor: E,
+  workspace_id: &Uuid,
+  namespace: &str,
+) -> Result<bool, AppError> {
+  let res = sqlx::query_scalar!(
+    r#"
+      SELECT EXISTS(
+        SELECT 1
+        FROM af_workspace
+        WHERE workspace_id = $1
+          AND publish_namespace = $2
+      )
+    "#,
+    workspace_id,
+    namespace,
+  )
+  .fetch_one(executor)
+  .await?;
+
+  Ok(res.unwrap_or(false))
+}
+
+#[inline]
+pub async fn update_workspace_publish_namespace<'a, E: Executor<'a, Database = Postgres>>(
+  executor: E,
+  workspace_id: &Uuid,
+  new_namespace: &str,
+) -> Result<(), AppError> {
+  let res = sqlx::query!(
+    r#"
+      UPDATE af_workspace
+      SET publish_namespace = $1
+      WHERE workspace_id = $2
+    "#,
+    new_namespace,
+    workspace_id,
+  )
+  .execute(executor)
+  .await?;
+
+  if res.rows_affected() != 1 {
+    tracing::error!(
+      "Failed to update workspace publish namespace, workspace_id: {}, new_namespace: {}, rows_affected: {}",
+      workspace_id, new_namespace, res.rows_affected()
+    );
+  }
+
+  Ok(())
+}
+
+#[inline]
+pub async fn select_workspace_publish_namespace<'a, E: Executor<'a, Database = Postgres>>(
+  executor: E,
+  workspace_id: &Uuid,
+) -> Result<String, AppError> {
+  let res = sqlx::query_scalar!(
+    r#"
+      SELECT publish_namespace
+      FROM af_workspace
+      WHERE workspace_id = $1
+    "#,
+    workspace_id,
+  )
+  .fetch_one(executor)
+  .await?;
+
+  Ok(res)
+}
+
+#[inline]
+pub async fn insert_or_replace_publish_collabs<'a, E: Executor<'a, Database = Postgres>>(
```

Comment:
Generally speaking this might be a very heavy query. We know that individual collabs can have MBs in size. If there's no upper bound on their size/count, we might have as well lock the table for undefined time.

Reply:
there is an upper bound 4M for metadata, 128MB for blob. that was check before the request came in `post_publish_collabs_handler`. do you think these upper bounds are reasonable?

Reply:
Tbh. I think that 128MB could be used to DDoS us. But we can adjust it accordingly in the future.

Reply:
There's also a max limit of 256MB for a single request in our load balancer on this particular endpoint. in additional, this also consumes workspace storage space, so it would be a matter of tries before the user's limit is reached and will be blocked.